prow/crier/reporters/github/reporter.go

```diff
@@ -162,8 +165,61 @@ func (c *Client) Report(ctx context.Context, log *logrus.Entry, pj *v1.ProwJob)
 		defer lock.Release(1)
 	}
 
+	pjsToReport := func() []v1.ProwJob {
+		if len(pj.Spec.Refs.Pulls) != 1 {
+			return nil
+		}
+		// find all prowjobs from this PR
+		selector := map[string]string{}
+		for _, l := range []string{kube.OrgLabel, kube.RepoLabel, kube.PullLabel} {
+			selector[l] = pj.ObjectMeta.Labels[l]
+		}
+		var pjs v1.ProwJobList
+		if err := c.lister.List(ctx, &pjs, ctrlruntimeclient.MatchingLabels(selector)); err != nil {
+			log.WithError(err).Errorf("Cannot list prowjob with selector %v", selector)
+			return nil
+		}
+
+		latestBatch := make(map[string]v1.ProwJob)
+		for _, pjob := range pjs.Items {
+			if len(pjob.Spec.Refs.Pulls) != 1 {
+				continue
+			}
+			if pj.Spec.Refs.Pulls[0].SHA != pjob.Spec.Refs.Pulls[0].SHA {
+				continue
+			}
+			// Now you have convinced me that you are the same job from my revision,
+			// continue convince me that you are the last one of your kind
+			if existing, ok := latestBatch[pjob.Spec.Job]; !ok {
+				latestBatch[pjob.Spec.Job] = pjob
+			} else if pjob.CreationTimestamp.After(existing.CreationTimestamp.Time) {
+				latestBatch[pjob.Spec.Job] = pjob
+			}
+		}
+
+		// If current pj is not the latest, we should skip
+		if existing, ok := latestBatch[pj.Spec.Job]; !ok {
+			// This should not happen unless this job was suddenly deleted right before we are trying to report
+			log.WithFields(logrus.Fields{"name": pj.Spec.Job, "job": pj.Name}).Error("Prowjob to be reported was not found")
+			return nil
+		} else if existing.Name != pj.Name {
+			return nil
+		}
+
+		// Check all jobs to make sure they are completed
+		var toReport []v1.ProwJob
+		for _, pjob := range latestBatch {
+			if !pjob.Complete() {
```

Comment:
Small performance nit, but we can do this check and the return already above when iterating over the jobs we got from the cache and save ourselves a bunch of map entries and checking or not?

Reply:
Moved up